cumulusci/tasks/bulkdata/utils.py

```diff
@@ -37,13 +39,27 @@ def _sql_bulk_insert_from_records(
         self, *, connection, table, columns, record_iterable
     ):
         """Persist records from the given generator into the local database."""
-        table = self.metadata.tables[table]
-
-        connection.execute(
-            table.insert(), [dict(zip(columns, row)) for row in record_iterable]
+        _consume(
+            self._sql_bulk_insert_from_records_incremental(
+                connection=connection,
+                table=table,
+                columns=columns,
+                record_iterable=record_iterable,
+            )
         )
 
-        self.session.flush()
+    def _sql_bulk_insert_from_records_incremental(
+        self, *, connection, table, columns, record_iterable
+    ):
+        """Generator that persists batches of records from the given generator into the local database
+
+        Yields after every batch."""
+        table = self.metadata.tables[table]
+        dict_iterable = (dict(zip(columns, row)) for row in record_iterable)
+        for group in _grouper(10000, dict_iterable):
+            with connection.begin():
```

Comment:
So this is creating a nested transaction for each group? What's the reason for that?

Reply:
I'm not totally sure what's going on either; my comment was based on the fact that https://docs.sqlalchemy.org/en/13/core/connections.html#sqlalchemy.engine.Connection.begin says it creates nested transactions when called repeatedly on the same connection, and I see it happening in a loop here without an obvious place where it gets committed.

Reply:
It's not repeated calls that create nested transactions but "nested calls". In this context I'm using the context manager syntax so only a nested context manager would create a nested transaction.

Ending the context of the context manager will commit. I confirmed that here:

https://github.com/zzzeek/sqlalchemy/blob/91100022ecf99ba0f1888e5a804b3d2d2bc2e128/lib/sqlalchemy/engine/base.py#L2131

Reply:
I see now - thanks.